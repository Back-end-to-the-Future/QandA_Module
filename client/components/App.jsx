import React from 'react';
import axios from 'axios';
import "regenerator-runtime/runtime.js";


import Question from './Question.jsx';
import SubmitQuestion from './SubmitQuestion.jsx';
import Search from './Search.jsx';

import Accordion from 'react-bootstrap/Accordion';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import InputGroup from 'react-bootstrap/InputGroup';
import FormControl from 'react-bootstrap/FormControl';


class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // props.questions is for testing purposes
      questions: this.props.questions,
      product_name: this.props.product_name,
      product_id: this.props.product_id,
      showSubmitQuestion: false,
      filterDisplay: false,
      filteredQuestions: []
    }

    this.getQuestions = this.getQuestions.bind(this);
    this.handleOpenSubmit = this.handleOpenSubmit.bind(this);
    this.handleCloseSubmit = this.handleCloseSubmit.bind(this);
    this.changeAnswers = this.changeAnswers.bind(this);
  }

  componentDidMount() {
    this.getQuestions();
  }

  async getQuestions(qLimit = 2, aLimit = 2) {
    let response = await axios.get('/questions', {
      params: {
        qLimit: qLimit,
        aLimit: aLimit,
        product_id: this.state.product_id
      }
    })

    let isMoreQuestions = response.data.isMoreQuestions;
    let questions = response.data.questions.sort((a, b) => (a.question_helpfulness > b.question_helpfulness) ? -1 : 1);
    this.setState({ questions, isMoreQuestions });
  }

  changeAnswers(question_id, newAnswers) {
    // makes deep copy of questions
    let questions = JSON.parse(JSON.stringify(this.state.questions));

    for (let i = 0; i < questions.length; i++) {
      if (questions[i].question_id === question_id) {
        questions[i].answers = newAnswers;
      }
    }

    this.setState({ questions });
  }

  handleOpenSubmit() {
    this.setState({ showSubmitQuestion: true });
  }

  handleCloseSubmit() {
    this.setState({ showSubmitQuestion: false });
  }

  render() {

    // handles undefined props.questions
    let displayQuestions = <span></span>;
    if (this.state.questions) {
      let questions = this.state.questions;
      if (this.state.filterDisplay) {
        questions = this.state.filteredQuestions;
      }

      displayQuestions = questions.map((question, index) => {
        return (
          <Question product_name={this.props.product_name}
            changeAnswers={this.changeAnswers}
            question={question} key={index}
          />
        );
      })
    }

    let seeMoreQuestions = <span></span>;
    if (this.state.isMoreQuestions) {
      seeMoreQuestions = <Button variant='outline-dark' className='moreQsButton'
        onClick={() => this.getQuestions(this.state.questions.length + 2)} >More Answered Questions</Button>
    }

    return (
      <div className='qaApp'>
        <Accordion className='qaAppInner'>
          <Card>
            <Card.Header>
              <Accordion.Toggle as={Button} variant="link" eventKey="0">
                Questions and Answers
              </Accordion.Toggle>
            </Card.Header>
            <Accordion.Collapse eventKey="0">
              <Card.Body>
                <Search
                  questions={this.state.questions}
                  setFilterState={(bool) => this.setState({ filterDisplay: bool })}
                  setFilteredQuestions={(fQs) => this.setState({ filteredQuestions: fQs })}
                />

                <div className='qaDisplay'>
                  {displayQuestions}
                  <div className='seeMoreQs'>
                    {seeMoreQuestions}
                  </div>
                </div>

                <SubmitQuestion
                  handleOpenSubmit={this.handleOpenSubmit}
                  handleCloseSubmit={this.handleCloseSubmit}
                  product_name={this.state.product_name}
                  product_id={this.state.product_id}
                  show={this.state.showSubmitQuestion}
                />

              </Card.Body>
            </Accordion.Collapse>
          </Card>
        </Accordion>
      </div>
    );
  }

}

export default App;