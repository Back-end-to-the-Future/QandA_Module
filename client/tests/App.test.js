import React from 'react';
import App from '../components/App.jsx';
import {shallow} from 'enzyme';

describe('App Component', () => {
  it('starts with a count of 0', () => {
    const wrapper = shallow(<App />);
    const text = wrapper.find('.count').text();

    expect(text).toEqual('Current Count: 0');
  })

  it('increments the count when the button is clicked', () => {
    const wrapper = shallow(<App />);

  })
});