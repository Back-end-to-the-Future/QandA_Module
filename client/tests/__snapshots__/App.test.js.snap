// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match test snapshot 1`] = `
<div
  className="qaApp"
>
  <div
    className="qaAppInner accordion"
  >
    <div
      className="card"
    >
      <div
        className="accHeader card-header"
        onClick={[Function]}
      >
        Questions and Answers
        <svg
          fill="currentColor"
          height="20px"
          viewBox="0 0 1792 1792"
          width="20px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1683 808l-742 741q-19 19-45 19t-45-19l-742-741q-19-19-19-45.5t19-45.5l166-165q19-19 45-19t45 19l531 531 531-531q19-19 45-19t45 19l166 165q19 19 19 45.5t-19 45.5z"
          />
        </svg>
      </div>
      <div
        aria-expanded={null}
        className="collapse"
      >
        <div
          className="card-body"
        >
          <div
            className="searchBar"
          >
            <div
              className="mb-3 input-group"
            >
              <input
                className="searchInput form-control"
                onChange={[Function]}
                placeholder="Have a question? Search for answersâ€¦"
                type="text"
              />
            </div>
          </div>
          <div
            className="qaDisplay"
          >
            <div
              className="questionDiv"
            >
              <div
                className="question container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-auto"
                  >
                    <h5>
                      Q: 
                      Can I wash it?
                    </h5>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="qButtonRow col"
                  >
                    <span>
                      Helpful? 
                    </span>
                     
                    <button
                      className="qHelpfulBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Yes (
                      23
                      )
                    </button>
                    <button
                      className="qReportBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Report
                    </button>
                  </div>
                </div>
                <div
                  className="answerRow row"
                >
                  <div
                    className="answerCol col"
                  >
                    <div
                      className="answerContainer"
                    >
                      <span />
                      <div
                        className="answer container"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="answerBody col"
                          >
                            A: 
                            I wouldn't machine wash it
                          </div>
                        </div>
                        <div
                          className="row"
                        >
                          <div
                            className="col"
                          >
                            by 
                            skilover
                            , 
                            November 04, 2017
                          </div>
                        </div>
                        <div
                          className="answerThumbnailDiv row"
                        />
                        <div
                          className="row"
                        >
                          <div
                            className="aButtonCol col"
                          >
                            <span>
                              Helpful? 
                            </span>
                            <button
                              className="aHelpfulBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Yes (
                              6
                              )
                            </button>
                            <button
                              className="aReportBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Report
                            </button>
                          </div>
                        </div>
                      </div>
                      <div
                        className="answer container"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="answerBody col"
                          >
                            A: 
                            Only if you want to ruin it!
                          </div>
                        </div>
                        <div
                          className="row"
                        >
                          <div
                            className="col"
                          >
                            by 
                            skilover
                            , 
                            November 04, 2017
                          </div>
                        </div>
                        <div
                          className="answerThumbnailDiv row"
                        />
                        <div
                          className="row"
                        >
                          <div
                            className="aButtonCol col"
                          >
                            <span>
                              Helpful? 
                            </span>
                            <button
                              className="aHelpfulBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Yes (
                              5
                              )
                            </button>
                            <button
                              className="aReportBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Report
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="moreAnswersRow row"
                >
                  <button
                    className="moreAnswers btn btn-outline-dark"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    See More Answers
                  </button>
                </div>
                <div
                  className="submitAnswerRow row"
                >
                  <div
                    className="submitAnswer"
                  >
                    <div
                      className="submitAButton"
                    >
                      <button
                        className="btn btn-outline-dark"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Submit Answer
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="questionDiv"
            >
              <div
                className="question container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-auto"
                  >
                    <h5>
                      Q: 
                      How long does it last?
                    </h5>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="qButtonRow col"
                  >
                    <span>
                      Helpful? 
                    </span>
                     
                    <button
                      className="qHelpfulBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Yes (
                      5
                      )
                    </button>
                    <button
                      className="qReportBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Report
                    </button>
                  </div>
                </div>
                <div
                  className="answerRow row"
                >
                  <div
                    className="answerCol col"
                  >
                    <div
                      className="answerContainer"
                    >
                      <span />
                      <div
                        className="answer container"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="answerBody col"
                          >
                            A: 
                            Some of the seams started splitting the first time I wore it!
                          </div>
                        </div>
                        <div
                          className="row"
                        >
                          <div
                            className="col"
                          >
                            by 
                            sillyguy
                            , 
                            November 28, 2019
                          </div>
                        </div>
                        <div
                          className="answerThumbnailDiv row"
                        />
                        <div
                          className="row"
                        >
                          <div
                            className="aButtonCol col"
                          >
                            <span>
                              Helpful? 
                            </span>
                            <button
                              className="aHelpfulBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Yes (
                              6
                              )
                            </button>
                            <button
                              className="aReportBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Report
                            </button>
                          </div>
                        </div>
                      </div>
                      <div
                        className="answer container"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="answerBody col"
                          >
                            A: 
                            Super Long Time
                          </div>
                        </div>
                        <div
                          className="row"
                        >
                          <div
                            className="col"
                          >
                            by 
                            JJ
                            , 
                            June 30, 2020
                          </div>
                        </div>
                        <div
                          className="answerThumbnailDiv row"
                        />
                        <div
                          className="row"
                        >
                          <div
                            className="aButtonCol col"
                          >
                            <span>
                              Helpful? 
                            </span>
                            <button
                              className="aHelpfulBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Yes (
                              0
                              )
                            </button>
                            <button
                              className="aReportBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Report
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="moreAnswersRow row"
                >
                  <button
                    className="moreAnswers btn btn-outline-dark"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    See More Answers
                  </button>
                </div>
                <div
                  className="submitAnswerRow row"
                >
                  <div
                    className="submitAnswer"
                  >
                    <div
                      className="submitAButton"
                    >
                      <button
                        className="btn btn-outline-dark"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Submit Answer
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="questionDiv"
            >
              <div
                className="question container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-auto"
                  >
                    <h5>
                      Q: 
                      Where is this product made?
                    </h5>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="qButtonRow col"
                  >
                    <span>
                      Helpful? 
                    </span>
                     
                    <button
                      className="qHelpfulBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Yes (
                      2
                      )
                    </button>
                    <button
                      className="qReportBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Report
                    </button>
                  </div>
                </div>
                <div
                  className="answerRow row"
                >
                  <div
                    className="answerCol col"
                  >
                    <div
                      className="answerContainer"
                    >
                      <span />
                      <div
                        className="answer container"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="answerBody col"
                          >
                            A: 
                            Canada
                          </div>
                        </div>
                        <div
                          className="row"
                        >
                          <div
                            className="col"
                          >
                            by 
                            footballfan
                            , 
                            August 06, 2018
                          </div>
                        </div>
                        <div
                          className="answerThumbnailDiv row"
                        />
                        <div
                          className="row"
                        >
                          <div
                            className="aButtonCol col"
                          >
                            <span>
                              Helpful? 
                            </span>
                            <button
                              className="aHelpfulBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Yes (
                              9
                              )
                            </button>
                            <button
                              className="aReportBtn btn btn-outline-dark btn-sm"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              Report
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="moreAnswersRow row"
                >
                  <span />
                </div>
                <div
                  className="submitAnswerRow row"
                >
                  <div
                    className="submitAnswer"
                  >
                    <div
                      className="submitAButton"
                    >
                      <button
                        className="btn btn-outline-dark"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Submit Answer
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="questionDiv"
            >
              <div
                className="question container"
              >
                <div
                  className="row"
                >
                  <div
                    className="col-auto"
                  >
                    <h5>
                      Q: 
                      What is the fabric made out of?
                    </h5>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="qButtonRow col"
                  >
                    <span>
                      Helpful? 
                    </span>
                     
                    <button
                      className="qHelpfulBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Yes (
                      1
                      )
                    </button>
                    <button
                      className="qReportBtn btn btn-outline-dark btn-sm"
                      disabled={false}
                      onClick={[Function]}
                      type="button"
                    >
                      Report
                    </button>
                  </div>
                </div>
                <div
                  className="answerRow row"
                >
                  <div
                    className="answerCol col"
                  >
                    <div
                      className="answerContainer"
                    >
                      <div
                        className="fade alert alert-secondary show"
                        id="noAnswerAlert"
                        role="alert"
                      >
                        I'm sorry, no one has answered this question yet.
                        <br />
                        If you have an answer please Submit it below.
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="moreAnswersRow row"
                >
                  <span />
                </div>
                <div
                  className="submitAnswerRow row"
                >
                  <div
                    className="submitAnswer"
                  >
                    <div
                      className="submitAButton"
                    >
                      <button
                        className="btn btn-outline-dark"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        Submit Answer
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="seeMoreQs"
            >
              <span />
            </div>
          </div>
          <div
            className="submitQuestion"
          >
            <div
              className="submitQButton"
            >
              <button
                className="btn btn-outline-dark"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Submit Question
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
